src/main/java/com/google/sps/data/Journey.java

```diff
@@ -56,12 +59,235 @@ public Pair(int first, int second) {
+    
+    int minTime = bestTime.get(currentState).get(0).first;
+    // currentWaypoint is the one for which bestTime[currentState][currentWaypoint] is minimal
+    int currentWaypoint = 0;
+    for (int waypointIndex = 1; waypointIndex < numberOfWaypoints; ++waypointIndex) {
+      if (bestTime.get(currentState).get(waypointIndex).first < minTime) {
+        minTime = bestTime.get(currentState).get(waypointIndex).first;
+        currentWaypoint = waypointIndex;
+      }
+    }
+    
+    if (minTime == Integer.MAX_VALUE) {
+      // No solution was found.
+      throw new BadRequestException("Impossible to create journey with the given constraints!");
+    }
+    
+    while (currentState != 0) {
+      orderedWaypoints.add(waypoints.get(currentWaypoint));
+      int prevState = currentState ^ (1 << currentWaypoint);
+      int prevWaypoint = bestTime.get(currentState).get(currentWaypoint).second;
+      currentWaypoint = prevWaypoint;
+      currentState = prevState;
+    }
+    // add the start of the journey at the end of orderedWaypoints.
+    orderedWaypoints.add(start);
+    Collections.reverse(orderedWaypoints);
+    return orderedWaypoints;
+  }
+
+  /**
+   * Creates and initializes bestTime matrix, a matrix where bestTime[config][lastWaypoint] = the
+   * minimum time to visit all points with bits set to 1 in config such that the last visited point
+   * is lastWaypoint; config is a bitmask where bits set to 1 represent currently visited waypoints.
+   */
+  private ArrayList<ArrayList<Pair>> initializeBestTimeMatrix() throws ApiException, IOException, InterruptedException, DataNotFoundException {
+    ArrayList<ArrayList<Pair>> bestTime = new ArrayList<>();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    for (int config = 0; config < (1 << numberOfWaypoints); ++config) {
+      // add the config-th line to the matrix with numberOfWaypoints columns of Integer.MAX_VALUE
+      ArrayList<Pair> currentRow = new ArrayList<Pair>();
+
+      for (int waypointIndex = 0; waypointIndex < numberOfWaypoints; ++ waypointIndex) {
+        currentRow.add(new Pair(Integer.MAX_VALUE, -1));
+      }
+      bestTime.add(currentRow);
+    }
+
+    int waypointIndex = 0;
+    for (CourierStop waypoint : waypoints) {
+      Pair currentElement = bestTime.get(1 << waypointIndex).get(waypointIndex);
+      currentElement.first = pathFinder.getTimeInSecondsBetweenPoints(start.getPoint(), waypoint.getPoint());
+      waypointIndex += 1;
+    }
+
+    return bestTime;
+  }
+
+  /**
+   * Computes and returns the bestTime matrix such that bestTime[config][latsWaypoint].first is the
+   * min time needed to visit all points with bits set to 1 in config such that the path ends in 
+   * lastWaypoint and bestTime[config][latsWaypoint].second represents the point that was visited 
+   * right before lastWaypoint. 
+   *
+   * The method will call initializeBestTimeMatrix() to set the initial states for bestTime.
```

Comment:
The fact that this method will call another is an implementation detail. You shouldn't document things like this in the contract of the method (and the user doesn't really care that this calls a private method anyway).

What you should document is behavior, which you also do (e.g. This will return a matrix where bla bla) and any caveats (e.g. it will reutrn (MAX_VALUE, -1) for row 0).

What happens if there is no route which respects the provided restrictions, what will each element be set to? Also (MAX_VALUE, -1)?